server/secure_test.go

```diff
@@ -0,0 +1,395 @@
+package server
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+
+	"github.com/theplant/appkit/log"
+)
+
+//////////////////////////////////////////////////
+// CSRF Required Header Verification
+
+func ExampleCSRFHeader_Blank() {
+	cfg := CrossSiteConfig{
+		RawAllowedOrigins:  "http://example.com",
+		CSRFRequiredHeader: "", // Explicitly blank
+	}
+
+	s, req := setup(cfg)
+
+	req.Header.Add("origin", "http://example.com")
+
+	exec(s, req)
+
+	// Output: executed handler
+	// 200
+}
+
+func ExampleCSRFHeader_Set() {
+	cfg := CrossSiteConfig{
+		RawAllowedOrigins:  "http://example.com",
+		CSRFRequiredHeader: "X-Csrf", // Same as default
+	}
+
+	s, req := setup(cfg)
+
+	req.Header.Add("origin", "http://example.com")
+	req.Header.Add("X-Csrf", ".") // Value is ignored
```

Comment:
I see the rails and asp.net implement use token. Not only validate the header if exists.

And they said from the API.

> If the request uses a safe HTTP method (GET, HEAD, OPTIONS, TRACE), the antiforgery token is not validated

https://github.com/aspnet/Antiforgery/blob/1c9a4f5705c02d182e6beefa96739de81cd3d850/src/Microsoft.AspNetCore.Antiforgery/IAntiforgery.cs

Maybe there are many implement.

Reply:
This middleware is designed to be used for split front-end and API. And tokens are not so simple to implement with that architecture.

So, I've implemented *Origin/Referrer verification* and *Use of Custom Request Headers*: https://www.owasp.org/index.php/Cross-Site_Request_Forgery_(CSRF)_Prevention_Cheat_Sheet

For apps that render HTML in the same server as the app, we can use CSRF tokens (something like https://github.com/justinas/nosurf).

Reply:
Thanks for your explanation.